### PR TITLE
[WL-1052] feat: swap history

### PR DESCRIPTION
**Pull request by FME849**

### CODE REVIEW

File: apps/wallet/src/engine/runners/solana/history/index.ts [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+		receivedTokenAmount,
+		sentTokenMint,
+		sentTokenAmount,
+	} = extractSwapBalance(transaction, meta);
+
+	if (!receivedTokenMint || !sentTokenMint)
+		return { transactionType: 'Unknown' };
+
+	const isWrappedSol =
+		receivedTokenMint === 'So11111111111111111111111111111111111111112';
```
Comment (tanlethanh) [inline apps/wallet/src/engine/runners/solana/history/index.ts:216]: you could create a `wrappedSolMint` constant in `utils/constants`

File: apps/wallet/src/engine/runners/solana/history/index.ts [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+	TokenBalance,
+} from '@solana/web3.js';
+import type { PublicKey } from '@solana/web3.js';
+import type {
+	SolanaSwapHistory,
+	SolanaSwapHistoryV2,
+	SolanaToken,
+	SolanaTransferHistory,
+	SolanaTransferHistoryV2,
+	SolanaUnknownHistory,
```
Comment (tanlethanh) [inline apps/wallet/src/engine/runners/solana/history/index.ts:16]: V2 will be the current used types, so you could deprecate the old types with `v1` prefix and remove `v2` of current types

File: apps/wallet/src/engine/runners/solana/history/index.ts [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+
+	return [...new Set(signatures)];
+};
+
+const constructTransactionHistoryDocument = async (
+	connection: Connection,
+	parsedTransaction: ParsedTransactionWithMeta,
+	wallet: PublicKey,
+): Promise<
+	TransactionHistoryDocument<
```
Comment (tanlethanh) [inline apps/wallet/src/engine/runners/solana/history/index.ts:109]: for short, you can use `HistoryDocument` and `SolanaTransfer`, `SolanaSwap`

File: apps/wallet/src/engine/runners/solana/history/index.ts [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+	const { meta, transaction, blockTime } = parsedTransaction;
+	const walletAddress = wallet.toString();
+
+	const _id = `${walletAddress}/history/${transaction.signatures[0]}`;
+	const signature = transaction.signatures[0];
+	const id = signature;
+	const date = blockTime ? new Date(blockTime * 1000) : new Date();
+	const status = !meta?.err ? 'Success' : 'Failed';
+	const tokenForFee = getTransactionTokenFee();
+	const fee = (meta?.fee || 0) / 10 ** 9;
```
Comment (tanlethanh) [inline apps/wallet/src/engine/runners/solana/history/index.ts:122]: could replace (10 ** 9) by LAMPORTS_PER_SOL

File: apps/wallet/src/engine/runners/solana/history/index.ts [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+	const balances = await getTransactionBalances(
+		connection,
+		parsedTransaction,
+		walletAddress,
+	);
+
+	if (!tokenForFee || balances.transactionType === 'Unknown') {
+		return {
+			_id,
+			id,
```
Comment (tanlethanh) [inline apps/wallet/src/engine/runners/solana/history/index.ts:132]: why do we need two id fields

File: apps/wallet/src/engine/runners/solana/history/index.ts [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+		storage.safeGet<TokenDocument<SolanaToken>>(sentTokenId),
+	]);
+
+	const receivedToken = {
+		amount: receivedTokenAmount / 10 ** (receivedTokenDoc?.decimals || 9),
+		metadata: {
+			mint: receivedTokenMint,
+			name: receivedTokenDoc?.name || '',
+			symbol: receivedTokenDoc?.symbol || '',
+			image: receivedTokenDoc?.image || '',
```
Comment (tanlethanh) [inline apps/wallet/src/engine/runners/solana/history/index.ts:234]: could use "Unknown" here for `name` and `symbol`, I applied this default for token and nft also

File: apps/wallet/src/engine/runners/solana/history/index.ts [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+		name: solMetadata.name,
+		symbol: solMetadata.symbol,
+		image: solMetadata.image,
+	};
+	const parsedInstruction = transaction.message.instructions.find(
+		(instruction) =>
+			'parsed' in instruction && instruction.parsed !== 'undefined',
+	) as ParsedInstruction;
+	const amount =
+		((parsedInstruction?.parsed?.info?.lamports as number) || 0) / 10 ** 9;
```
Comment (tanlethanh) [inline apps/wallet/src/engine/runners/solana/history/index.ts:279]: use LAMPORTS_PER_SOL

File: apps/wallet/src/engine/runners/solana/index.ts [hunk 8]
```diff
@@ -77,7 +77,7 @@ export const createSolanaRunner: CreateFunction = async (config) => {
 					}),
 					watchLogs(connection, cluster, wallet),
 					watchAccount(connection, cluster, wallet, wallet),
-					getTransactionsHistory(connection, cluster, wallet, accounts),
+					getTransactionsHistory(connection, wallet, accounts),
```
Comment (tanlethanh) [inline apps/wallet/src/engine/runners/solana/index.ts:80]: You might need to add cluster for the history, in the future, we might support both devnet and mainnet, so keep the cluster might be able to filter out tokens/nfts/history

File: apps/wallet/src/engine/runners/solana/types.ts [hunk 9]
```diff
@@ -16,3 +23,28 @@ export type ParsedTokenAccount = {
...
+	postBalance?: TokenAmount;
+	preBalance?: TokenAmount;
+}
+
+export type AccountBalanceMap = Record<string, AccountBalance>;
+
+export type TransactionBalances =
+	| {
+			transactionType: 'Unknown';
+	  }
```
Comment (tanlethanh) [inline apps/wallet/src/engine/runners/solana/types.ts:37]: could create a new type for this one

**Review by tanlethanh — APPROVED** [review]

File: apps/wallet/src/engine/runners/solana/history/index.ts (hunk 5 above)
Comment (FME849) [reply apps/wallet/src/engine/runners/solana/history/index.ts:132]: just a legacy shape of history document, I removed it

**Review by tanlethanh — APPROVED** [review]

**Pull request merged**